### PR TITLE
ICU-20183 Adding CurrencyUnit StringPiece constructor.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20183
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
I found a bug involving NUL terminators and pushed another commit to fix it.  Please check.  If it looks good, I will squash the commits.

File: icu4c/source/i18n/currunit.cpp [hunk 1]
```diff
@@ -50,6 +52,30 @@ CurrencyUnit::CurrencyUnit(ConstChar16Ptr _isoCode, UErrorCode& ec) {
     initCurrency(simpleIsoCode);
 }
 
+CurrencyUnit::CurrencyUnit(StringPiece _isoCode, UErrorCode& ec) {
+    // Note: unlike the old constructor, reject empty arguments with an error.
+    char isoCodeBuffer[4];
+    const char* isoCodeToUse;
+    // uprv_memchr checks that the string contains no internal NULs
+    if (_isoCode.length() != 3 || uprv_memchr(_isoCode.data(), 0, 3) != nullptr) {
```
Comment (jefgen) [inline icu4c/source/i18n/currunit.cpp]: I don't think we have a `uprv_memchr` though. Perhaps we should add one?
(Assuming you mean a wrapper over [memchr](https://en.cppreference.com/w/c/string/byte/memchr)?)

File: icu4c/source/test/intltest/numfmtst.cpp [hunk 2]
```diff
@@ -2120,6 +2120,17 @@ void NumberFormatTest::TestCurrencyUnit(void){
     static const char XXX8[]  =  "XXX";
     static const UChar INV[]  = u"{$%";
     static const char INV8[]  =  "{$%";
+    static const UChar ZZZ[]  = u"zz";
```
Comment (jefgen) [inline icu4c/source/test/intltest/numfmtst.cpp]: Would it be maybe worth adding a BAD3 test for the embedded nul (U_ILLEGAL_ARGUMENT_ERROR)?

File: icu4c/source/i18n/currunit.cpp (hunk 1 above)
Comment (sffc) [reply icu4c/source/i18n/currunit.cpp]: Yeah, I added it but forgot to commit the file.  I just copied how we implement uprv_memcpy:

```cpp
#define uprv_memset(buffer, mark, size) U_STANDARD_CPP_NAMESPACE memset(buffer, mark, size)
#define uprv_memcmp(buffer1, buffer2, size) U_STANDARD_CPP_NAMESPACE memcmp(buffer1, buffer2,size)
#define uprv_memchr(ptr, value, num) U_STANDARD_CPP_NAMESPACE memchr(ptr, value, num)
```

Commit pushed.

File: icu4c/source/test/intltest/numfmtst.cpp (hunk 2 above)
Comment (sffc) [reply icu4c/source/test/intltest/numfmtst.cpp]: This test case already exercises that path:

```cpp
} else if (_isoCode[1] == 0 || _isoCode[2] == 0) {
```

Down below, for the 8-bit char string case, I run the constructor both with length 2 and with length 3 to exercise both branches of the 8-bit constructor.

Comment (jefgen) [reply icu4c/source/test/intltest/numfmtst.cpp]: Ah I missed this: `CurrencyUnit zzz8b({ZZZ8, 3}, ec); `

Thanks Shane!

**Comment by sffc** [comment]
Thanks for the LGTM.  Pushed a squash commit and need another one.

**Comment by sffc** [comment]
deptest says:

> Error: in group units: i18n/currunit.o imports memchr - is this a new system symbol?

I'll add that to dependencies.txt.  Ok @markusicu ?

**Comment by markusicu** [comment]
> deptest says:
> 
> > Error: in group units: i18n/currunit.o imports memchr - is this a new system symbol?
> 
> I'll add that to dependencies.txt. Ok @markusicu ?

Yes. Presumably to one of the system symbol groups

**Pull request merged**